### PR TITLE
mds: avoid the assert failure when the inode for the cap_export from other MDS happened not in MDCache

### PR DESCRIPTION
**Pull request by JianyuLi**
Fixes: http://tracker.ceph.com/issues/22610
Signed-off-by: Jianyu Li <joannyli@tencent.com>

### CODE REVIEW

**Comment by ukernel** [comment]
something was wrong

The inode should be loaded into cache during determining the cap export target. The crash mds was recovering too, it should not trim its cache.

**Comment by JianyuLi** [comment]
Hi Zheng,

Let me give some more details about this issue:
1. mds.1 was killed by OOM, then its standby-replay peer took the active role;
2. new mds.1 started the up:rejoin procedure;
3. mds.0  restarted due to unknown reason(its log has been rotated and deleted);
4.  the standby-replay peer of mds.0 took the active role and after it received the weak-rejoin request in step 2, it throw the assert, the log at that moment is as below:

2018-01-04 20:45:57.042072 7f2a0fc01700  1 mds.0.694709 resolve_done
2018-01-04 20:45:57.880314 7f2a0ebff700  1 mds.0.694709 handle_mds_map i am now mds.0.694709
2018-01-04 20:45:57.880320 7f2a0ebff700  1 mds.0.694709 handle_mds_map state change up:resolve --> up:reconnect
2018-01-04 20:45:57.880334 7f2a0ebff700  1 mds.0.694709 reconnect_start
2018-01-04 20:45:57.880527 7f2a0ebff700  1 mds.0.server reconnect_clients -- 434 sessions
2018-01-04 20:46:45.651618 7f2a0c3fa700  1 mds.0.694709 reconnect_done
2018-01-04 20:46:46.132713 7f2a0ebff700  1 mds.0.694709 handle_mds_map i am now mds.0.694709
2018-01-04 20:46:46.132719 7f2a0ebff700  1 mds.0.694709 handle_mds_map state change up:reconnect --> up:rejoin
2018-01-04 20:46:46.132735 7f2a0ebff700  1 mds.0.694709 rejoin_start
2018-01-04 20:46:46.134740 7f2a0ebff700  1 mds.0.cache process_imported_caps missing ino count: 37
2018-01-04 20:46:46.134749 7f2a0ebff700  1 mds.0.694709 rejoin_joint_start
2018-01-04 20:46:46.597072 7f2a0fc01700 -1 mds/MDCache.cc: In function 'void MDCache::handle_cache_rejoin_weak(MMDSCacheRejoin*)' thread 7f2a0fc01700 time 2018-01-04 20:46:46.593301
mds/MDCache.cc: 4325: FAILED assert(in && in->is_auth())
 ceph version 10.2.9-102-g820619c (820619cc59a3790ab36be1945a135eb826c558f1)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x85) [0x7f2a15885205]
 2: (MDCache::handle_cache_rejoin_weak(MMDSCacheRejoin*)+0x63e) [0x7f2a1558b59e]
 3: (MDCache::handle_cache_rejoin(MMDSCacheRejoin*)+0x25b) [0x7f2a1559104b]
...

You mentioned that the inode should be loaded into cache during cap export target. I think this is right for mds.1,  only in this way mds.1 was able to know itself isn't the auth of that inode and should forward it to mds.0 through [ in->authority().first ]. But I am afraid that this is the fact for mds.0, especially for the mds.0 changed from standby-replay mode  in our case.  It is highly likely that this inode haven't been existed in the mds.0 cache ever.  So there are multiple call of process_imported_caps function during rejoin phase, one is in rejoin_start which handles the imported caps from its own reconnect phase, the other in rejoin_gather_finish function which handles the imported cap forwarded by others. 

In fact, even through we have finished rejoin_gather, there is chance that some inodes in the cap_imports map couldn't be found in MDCache, which should be handled after client_replay phase.  We could see the related code in process_imported_caps:
bool MDCache::process_imported_caps()
{
...
  // called by rejoin_gather_finish() ?
  if (rejoin_gather.count(mds->get_nodeid()) == 0) {
   ...
   // process cap imports
    //  ino -> client -> frommds -> capex
    for (auto p = cap_imports.begin(); p != cap_imports.end(); ) {
      CInode *in = get_inode(p->first);
      if (!in) {
	dout(10) << " still missing ino " << p->first
	         << ", will try again after replayed client requests" << dendl;
	++p;
	continue;
      }
      assert(in->is_auth());
  ...
}

In conclusion, I think it had better to modify the assert logic because (in && in->is_auth()) is too strict.

**Comment by ukernel** [comment]
Thank you for your detailed explaination
